diff --git a/src/core/nm-config.c b/src/core/nm-config.c
index 878f343a54..cd3dc81df8 100644
--- a/src/core/nm-config.c
+++ b/src/core/nm-config.c
@@ -850,6 +850,7 @@ static const ConfigGroup config_groups[] = {
                              NM_CONFIG_KEYFILE_KEY_MAIN_DHCP,
                              NM_CONFIG_KEYFILE_KEY_MAIN_DNS,
                              NM_CONFIG_KEYFILE_KEY_MAIN_FIREWALL_BACKEND,
+                             NM_CONFIG_KEYFILE_KEY_MAIN_FIREWALL_BACKEND_PATH,
                              NM_CONFIG_KEYFILE_KEY_MAIN_HOSTNAME_MODE,
                              NM_CONFIG_KEYFILE_KEY_MAIN_IGNORE_CARRIER,
                              NM_CONFIG_KEYFILE_KEY_MAIN_IWD_CONFIG_PATH,
diff --git a/src/core/nm-firewall-utils.c b/src/core/nm-firewall-utils.c
index b6b2b0d721..9da942d8da 100644
--- a/src/core/nm-firewall-utils.c
+++ b/src/core/nm-firewall-utils.c
@@ -17,9 +17,9 @@
 
 /*****************************************************************************/
 
-static const struct {
-    const char *name;
-    const char *path;
+static struct {
+    char *name;
+    char *path;
 } FirewallBackends[] = {
     [NM_FIREWALL_BACKEND_NONE - 1] =
         {
@@ -213,7 +213,7 @@ _share_iptables_call_v(const char *const *argv)
 }
 
 #define _share_iptables_call(...) \
-    _share_iptables_call_v(NM_MAKE_STRV("" IPTABLES_PATH "", "--wait", "2", __VA_ARGS__))
+    _share_iptables_call_v(NM_MAKE_STRV(FirewallBackends[NM_FIREWALL_BACKEND_IPTABLES-1].path, "--wait", "2", __VA_ARGS__))
 
 static gboolean
 _share_iptables_chain_op(const char *table, const char *chain, const char *op)
@@ -598,7 +598,7 @@ nm_firewall_nft_call(GBytes             *stdin_buf,
     g_subprocess_launcher_set_environ(subprocess_launcher, NM_STRV_EMPTY());
 
     call_data->subprocess = g_subprocess_launcher_spawnv(subprocess_launcher,
-                                                         NM_MAKE_STRV(NFT_PATH, "-f", "-"),
+                                                         NM_MAKE_STRV(FirewallBackends[NM_FIREWALL_BACKEND_NFTABLES-1].path, "-f", "-"),
                                                          &error);
 
     if (!call_data->subprocess) {
@@ -1010,17 +1010,6 @@ nm_firewall_config_apply_sync(NMFirewallConfig *self, gboolean up)
 
 /*****************************************************************************/
 
-static NMFirewallBackend
-_firewall_backend_detect(void)
-{
-    if (g_file_test(NFT_PATH, G_FILE_TEST_IS_EXECUTABLE))
-        return NM_FIREWALL_BACKEND_NFTABLES;
-    if (g_file_test(IPTABLES_PATH, G_FILE_TEST_IS_EXECUTABLE))
-        return NM_FIREWALL_BACKEND_IPTABLES;
-
-    return NM_FIREWALL_BACKEND_NFTABLES;
-}
-
 NMFirewallBackend
 nm_firewall_utils_get_backend(void)
 {
@@ -1031,7 +1020,7 @@ again:
     b = g_atomic_int_get(&backend);
     if (b == NM_FIREWALL_BACKEND_UNKNOWN) {
         gs_free char *conf_value = NULL;
-        gboolean      detect;
+        gs_free char *path_value = NULL;
         int           i;
 
         conf_value =
@@ -1049,9 +1038,18 @@ again:
             }
         }
 
-        detect = (b == NM_FIREWALL_BACKEND_UNKNOWN);
-        if (detect)
-            b = _firewall_backend_detect();
+        path_value = 
+            nm_config_data_get_value(NM_CONFIG_GET_DATA_ORIG,
+                                     NM_CONFIG_KEYFILE_GROUP_MAIN,
+                                     NM_CONFIG_KEYFILE_KEY_MAIN_FIREWALL_BACKEND_PATH,
+                                     NM_CONFIG_GET_VALUE_STRIP | NM_CONFIG_GET_VALUE_NO_EMPTY);
+
+        if (path_value == NULL) {
+            nm_log_err(LOGD_SHARING, "firewall: backend defined without a path!");
+            g_atomic_int_set(&backend, NM_FIREWALL_BACKEND_UNKNOWN);
+        } else {
+            FirewallBackends[b].path = path_value;
+        }
 
         nm_assert(NM_IN_SET(b,
                             NM_FIREWALL_BACKEND_NONE,
@@ -1062,18 +1060,12 @@ again:
             goto again;
 
         nm_log_dbg(LOGD_SHARING,
-                   "firewall: use %s backend%s%s%s%s%s%s%s",
+                   "firewall: use %s backend%s%s%s",
                    FirewallBackends[b - 1].name,
                    NM_PRINT_FMT_QUOTED(FirewallBackends[b - 1].path,
                                        " (",
                                        FirewallBackends[b - 1].path,
                                        ")",
-                                       ""),
-                   detect ? " (detected)" : "",
-                   NM_PRINT_FMT_QUOTED(detect && conf_value,
-                                       " (invalid setting \"",
-                                       conf_value,
-                                       "\")",
                                        ""));
     }
 
diff --git a/src/libnm-base/nm-config-base.h b/src/libnm-base/nm-config-base.h
index 362a183c93..29cdee12c4 100644
--- a/src/libnm-base/nm-config-base.h
+++ b/src/libnm-base/nm-config-base.h
@@ -27,6 +27,7 @@
 #define NM_CONFIG_KEYFILE_KEY_MAIN_DHCP                        "dhcp"
 #define NM_CONFIG_KEYFILE_KEY_MAIN_DNS                         "dns"
 #define NM_CONFIG_KEYFILE_KEY_MAIN_FIREWALL_BACKEND            "firewall-backend"
+#define NM_CONFIG_KEYFILE_KEY_MAIN_FIREWALL_BACKEND_PATH       "firewall-backend-path"
 #define NM_CONFIG_KEYFILE_KEY_MAIN_HOSTNAME_MODE               "hostname-mode"
 #define NM_CONFIG_KEYFILE_KEY_MAIN_IGNORE_CARRIER              "ignore-carrier"
 #define NM_CONFIG_KEYFILE_KEY_MAIN_IWD_CONFIG_PATH             "iwd-config-path"
@@ -37,6 +38,8 @@
 #define NM_CONFIG_KEYFILE_KEY_MAIN_RC_MANAGER                  "rc-manager"
 #define NM_CONFIG_KEYFILE_KEY_MAIN_SYSTEMD_RESOLVED            "systemd-resolved"
 
+// #define NM_CONFIG_KEYFILE_KEY_BIN_PATHS_IPTABLES               "iptables"
+
 #define NM_CONFIG_KEYFILE_KEY_LOGGING_AUDIT   "audit"
 #define NM_CONFIG_KEYFILE_KEY_LOGGING_BACKEND "backend"
 #define NM_CONFIG_KEYFILE_KEY_LOGGING_DOMAINS "domains"
